Studio: keymap subsystem (#2268)
zmkfirmware/zmk: docs/docs/features/studio.md

## Building

Building for ZMK Studio involves two main additional items.

- Build with the `studio-rpc-usb-uart` snippet to enable the endpoint used for ZMK Studio communication over USB.
- Enable the `ZMK_STUDIO` Kconfig setting.

### GitHub Actions

First add a `studio-rpc-usb-uart` to the `snippet` property of your build configuration. For a split keyboard, you should do this _only_ for your central/left side, e.g.:

```
---
include:
  - board: nice_nano_v2
    shield: corne_left
    snippet: studio-rpc-usb-uart
  - board: nice_nano_v2
    shield: corne_right
```

Next, enable the `ZMK_STUDIO` Kconfig symbol, for example by adding the following line to your .conf file:

```
CONFIG_ZMK_STUDIO=y
```

### Local Build

When building locally, use the `-S` parameter to include the `studio-rpc-usb-uart` snippet. Instead of adding it to your config file, you can also append the `ZMK_STUDIO` Kconfig as an additional CMake argument, e.g.:

```bash
west build -d build/cl_studio -b nice_nano_v2 \
  -S studio-rpc-usb-uart -- -DSHIELD=corne_left -DCONFIG_ZMK_STUDIO=y
```

## Including Unreferenced Behaviors

> Normally, ZMK will only build and include behaviors that are referenced by your keymap and unused behavior will be skipped. However, ZMK Studio builds using the `studio-rpc-usb-uart` snippet will automatically define the `ZMK_BEHAVIORS_KEEP_ALL` value, which changes the approach and builds all possible behaviors into the firmware. This lets those behaviors be used in ZMK Studio.

Reviewer: Note that there would be a name collision if someone were to create a behavior named "all". Not sure if that's likely enough to be worth changing this, but we could maybe add a word like "studio" somewhere before "keep" so it's in a different namespace?

Author: So... I'd originally used `STUDIO` in the naming, but in reality this isn't *really* a Studio specific thing, but a general way to keep behaviors in the DTS to ensure they're built, so I'm a bit resistant to re-introducing it.  I'm not sure it's *that* huge of a deal to have "ALL" be reserved, especially since the naming isn't *strictly* required to match 1-to-1.

A few controls are available to override this behavior for fine-grained control of what behaviors are built. Behaviors can be kept or omitted by defining certain values in the top of your keymap file before the standard `behaviors.dtsi` file is included.

### Omit Specific Behaviors

> You can omit a specific behaviors by defining a variable like `ZMK_BEHAVIORS_OMIT_FOO` at the top of your keymap:

Reviewer: "variable like `ZMK_BEHAVIORS_OMIT_FOO`" isn't super clear. Continuing the discussion above, we should have some description of how to map a behavior to a symbol to use in place of `FOO`.

Also, I think it is important (possibly even big warning box important) to note that these defines need to come *before* including behaviors.dtsi.

Author: I added a note to this, and there *is* a note about the order needed here, but I can make it even more obvious if folks thinks this needs it.

```dts
#define ZMK_BEHAVIORS_OMIT_KEY_TOGGLE

#include <behaviors.dtsi>
```

### Keep Only Selective Behaviors

> To override the default "keep all" functionality, you can define the `ZMK_BEHAVIORS_KEEP_SELECTIVE` flag, and then keep only specific behaviors with a flag like `ZMK_BEHAVIORS_KEEP_FOO`, e.g.:

Reviewer: Same deal as above with a name collision if someone were to name a behavior "selective". Also, do we actually need a separate symbol for this, or can we just
```c
#undef ZMK_BEHAVIORS_KEEP_ALL

#define ZMK_BEHAVIORS_KEEP_FOO
#define ZMK_BEHAVIORS_KEEP_BAR
```
instead?

Author: Let me *try* the `#undef` trick. I'm not sure if it'll work for symbols defined by passing them to the build command like is done with our snippet or not.

Author: Yeah, I like the `#undef` much better than the complex keep all/selective stuff. Works fine too, even when the symbol is originally defined by the snippet setting the `DTS_EXTRA_CPPFLAGS` value.

Author: Ok, fixed this up, including the documentation.

---
title: ZMK Studio
sidebar_label: Studio
---

:::warning[Alpha Feature]

> ZMK Studio support is in alpha at this point. Although our best efforts are being made to keep compatibility during active development, but is not guaranteed.

Reviewer 2: ```suggestion
ZMK Studio support is in alpha. Although best efforts are being made, keeping compatibility during active development is not guaranteed.
```

:::

> ZMK Studio provides runtime update functionality to ZMK powered devices. Studio is still under active development, and is not yet ready for casual end user use.

Reviewer 3: It might be nice to "dumb down" `runtime update functionality` a bit. Also, though not necessary, this could be a nice place to put an image or short video of ZMK Studio in use.

Author: Expanded on this just a bit.

## Building

Building for ZMK Studio involves two main additional items.

- Build with the `studio-rpc-usb-uart` snippet to enable the endpoint used for Studio communication over USB.
- Enable the `ZMK_STUDIO` Kconfig setting.

### GitHub Actions

First add a `studio-rpc-usb-uart` to the `snippet` property of your build configuration. For a split keyboard, you should do this _only_ for your central/left side, e.g.:

```
---
include:
  - board: nice_nano_v2
    shield: corne_left
    snippet: studio-rpc-usb-uart
  - board: nice_nano_v2
    shield: corne_right
```

Next, enable the `ZMK_STUDIO` Kconfig symbol, for example by adding the following line to your .conf file:

```
CONFIG_ZMK_STUDIO=y
```

### Local Build

> When building locally, use the `-S` parameter to include the `studio-rpc-usb-uart` snippet, and append the `ZMK_STUDIO` Kconfig as an additional CMake argument, e.g.:

Reviewer 2: ```suggestion
When building locally, use the `-S` parameter to include the `studio-rpc-usb-uart` snippet. Instead of adding it to your config file, you can also append the `ZMK_STUDIO` Kconfig as an additional CMake argument, e.g.:
```

```bash
west build -d build/cl_studio -b nice_nano_v2 \
  -S studio-rpc-usb-uart -- -DSHIELD=corne_left -DCONFIG_ZMK_STUDIO=y
```

> ## Behaviors

Reviewer 3: ```suggestion
## Including Unused Behaviors
```
I think it would be good to make the "unused" part explicit in the header for people who skim.

Author: Updated, but used the "Unreferenced" terminology instead.

> Normally, ZMK will only build and include behaviors that are referenced by your keymap, and unused behavior will be skipped. However, studio builds using the `studio-rpc-usb-uart` snippet will automatically define the `ZMK_BEHAVIORS_KEEP_ALL` value, which changes the approach and builds all possible behaviors into the firmware.

Reviewer 2: ```suggestion
Normally, ZMK will only build and include behaviors that are referenced by your keymap and unused behavior will be skipped. However, Studio builds using the `studio-rpc-usb-uart` snippet will automatically define the `ZMK_BEHAVIORS_KEEP_ALL` value, which changes the approach and builds all possible behaviors into the firmware. This allows those behaviors to be used in the Studio.
```

> A few controls are available to override this behavior for fine-grained control of what behaviors are built. Behaviors can be kept or omitted by defining certain values in the top of your keymap file before the standard `behaviors.dtsi` file is included.

Reviewer 2: #I think we need to provide a reference for what defines can be used. If we don't want to pollute here, we can just drop a link to a header file?

Author: They aren't defined by default. They only exists if the user defines them, which triggers the omit/keep functionality.

Reviewer 2: What I mean is that the user needs some way to figure out what `name` in `ZMK_BEHAVIORS_OMIT_##_name` corresponds to for a given behavior.

Author: Oh, fully agreed! We just sadly can't point to a header for this. Still feels like this should be either on each behavior page, or maybe in the behaviors index?

Reviewer 2: Right, there doesn't seem to be a single file where that can be checked. Do you think we could simply use the behavior reference to avoid adding a new naming convention? e.g. 
`ZMK_BEHAVIORS_OMIT_mt` for `&mt`. (I think partial lower case might be acceptable given the tradeoffs.)

Author: Let me ponder that one... I have mixed feelings on that on. Just a convention of "capitalize the normal reference" seems easy enough.

### Omit Specific Behaviors

You can omit a specific behaviors by defining a variable like `ZMK_BEHAVIORS_OMIT_FOO` at the top of your keymap:

```dts
#define ZMK_BEHAVIORS_OMIT_KEY_TOGGLE

#include <behaviors.dtsi>
```

### Keep Only Selective Behaviors

To override the default "keep all" functionality, you can define the `ZMK_BEHAVIORS_KEEP_SELECTIVE` flag, and then keep only specific behaviors with a flag like `ZMK_BEHAVIORS_KEEP_FOO`, e.g.:

```dts
#define ZMK_BEHAVIORS_KEEP_SELECTIVE

#define ZMK_BEHAVIORS_KEEP_STICKY_KEY
#define ZMK_BEHAVIORS_KEEP_MOD_TAP

#include <behaviors.dtsi>
```

> ## Layers

Reviewer 3: ```suggestion
## Extra Layers
```

Author: Added.

By default, a build with ZMK Studio enabled will only allow as many layers as are defined in your standard keymap. To make additional layers available for use through Studio, you simply add new empty layers to your keymap with a status of `reserved`, e.g.:

```dts
/ {
    keymap {
        compatible = "zmk,keymap";

        base {
            display-name = "Base";
            bindings = // etc.
        };

        fn_layer {
            display-name = "Fn";
            bindings = // etc.
        };

        extra1 {
            status = "reserved";
        };

>         extra2 {

Reviewer 3: If the names here are visible but unable to be changed in ZMK Studio, it might be nice to make a note of that here.

Author: Nope, names of layers are editable in ZMK Studio.

            status = "reserved";
        };
    }
}
```

> The reserved layers will be ignored during regular ZMK builds, but come available for ZMK Studio enabled builds.

Reviewer 2: ```suggestion
The reserved layers will be ignored during regular ZMK builds but will become available for ZMK Studio enabled builds.
```

Reviewer 2: I might have misinterpreted the original intention, feel free to ignore.

```

Next, enable the `ZMK_STUDIO` Kconfig symbol, for example by adding the following line to your .conf file:

```
CONFIG_ZMK_STUDIO=y
```

### Local Build

When building locally, use the `-S` parameter to include the `studio-rpc-usb-uart` snippet. Instead of adding it to your config file, you can also append the `ZMK_STUDIO` Kconfig as an additional CMake argument, e.g.:

```bash
west build -d build/cl_studio -b nice_nano_v2 \
  -S studio-rpc-usb-uart -- -DSHIELD=corne_left -DCONFIG_ZMK_STUDIO=y
```

## Including Unreferenced Behaviors

Normally, ZMK will only build and include behaviors that are referenced by your keymap and unused behavior will be skipped. However, ZMK Studio builds using the `studio-rpc-usb-uart` snippet will automatically define the `ZMK_BEHAVIORS_KEEP_ALL` value, which changes the approach and builds all possible behaviors into the firmware. This lets those behaviors be used in ZMK Studio.

A few controls are available to override this behavior for fine-grained control of what behaviors are built. Behaviors can be kept or omitted by defining certain values in the top of your keymap file, _before_ the standard `behaviors.dtsi` file is included.

> By convention, the defines used to keep/omit a given behavior are based on the behavior label, e.g. for the `&kt` behavior, the suffix to use would be `_KT`.

Reviewer 2: ```suggestion
By convention, the defines used to keep/omit a given behavior are typically upper-cased versions of the behavior label, e.g. for the `&kt` behavior, the suffix to use would be `_KT`.
```
Can we be more specific here?
